### PR TITLE
New Loader code based on upstream 1.3.271 compared against upstream

### PR DESCRIPTION
**Pull request by aqnuep**
NOTE: This PR is for review purposes only as this branch will replace `sc_main`, and not expected to be merged into `main` (upstream baseline code).

### CODE REVIEW

**Review by dgkoch — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 1]
```diff
@@ -16,13 +18,31 @@
...
+    add_definitions(-DVULKANSC)
+endif()
 
 add_subdirectory(scripts)
 
 set(THREADS_PREFER_PTHREAD_FLAG ON)
 find_package(Threads REQUIRED)
 
+# Add support for VULKANSC mode from the command line
+set(VULKANSC OFF CACHE BOOL "User defined variable for VULKANSC mode to be passed in through cmake command line e.g. -DVULKANSC=ON")
```
[the remark this answers is not in the document]
Comment (aqnuep) [reply CMakeLists.txt:44]: Sure. It does not make an effective difference as of now (CMake is a weird mixture of declarative and imperative languages), but better safe than sorry.

File: CMakeLists.txt [hunk 2]
```diff
@@ -95,11 +115,22 @@ endif()
...
+    if(NOT VULKANSC)
+        target_compile_definitions(platform_wsi INTERFACE VK_USE_PLATFORM_MACOS_MVK VK_USE_PLATFORM_METAL_EXT)
+    endif()
+elseif(UNIX AND NOT APPLE AND VULKANSC)
+    option(BUILD_WSI_XCB_SUPPORT "Build XCB WSI support" OFF)
+    option(BUILD_WSI_XLIB_SUPPORT "Build Xlib WSI support" OFF)
+    option(BUILD_WSI_WAYLAND_SUPPORT "Build Wayland WSI support" OFF)
+    option(BUILD_WSI_DIRECTFB_SUPPORT "Build DirectFB WSI support" OFF)
+    option(BUILD_WSI_SCREEN_QNX_SUPPORT "Build QNX Screen WSI support" OFF)
+    option(BUILD_WSI_SCI_SUPPORT "Build SCI WSI support" OFF)
```
[the remark this answers is not in the document]
Comment (aqnuep) [reply CMakeLists.txt:133]: Good catch! I'm not sure how I missed that.

File: loader/loader.c [hunk 3]
```diff
@@ -4691,18 +4710,18 @@ VkResult loader_create_instance_chain(const VkInstanceCreateInfo *pCreateInfo, c
...
 
-        for (uint32_t i = 0; i < pCreateInfo->enabledLayerCount; ++i) {
-            res = copy_str_to_string_list(inst, &inst->enabled_layer_names, pCreateInfo->ppEnabledLayerNames[i],
-                                          strlen(pCreateInfo->ppEnabledLayerNames[i]));
-            if (res != VK_SUCCESS) return res;
+            for (uint32_t i = 0; i < pCreateInfo->enabledLayerCount; ++i) {
+                res = copy_str_to_string_list(inst, &inst->enabled_layer_names, pCreateInfo->ppEnabledLayerNames[i],
+                                              strlen(pCreateInfo->ppEnabledLayerNames[i]));
+                if (res != VK_SUCCESS) return res;
+            }
```
[the remark this answers is not in the document]
Comment (aqnuep) [reply loader/loader.c:4724]: Oh yes!

This was the loader bug which caused `vkCreateInstance` to always return `VK_SUCCESS` if the application passes in any extensions in the create info, even if things actually failed.

This fix has already been upstreamed and should "come back" at the next downstreaming.

**Comment by aqnuep** [comment]
I've downstreamed 1.3.271, which introduced some changes to QNX builds as upstream now uses dedicated makefiles for the purpose, but I've kept CMake support as far as possible downstream.

**Comment by aqnuep** [comment]
Closing this PR now that the code in `sc_main` subsumes it. @sricker-nvidia, please use the code in `sc_main` for IHV testing purposes.

**Pull request closed without merging**